lib/trento/infrastructure/commanded/event_handlers/saptune_update_event_handler.ex

```diff
@@ -0,0 +1,210 @@
+  database as they might not be explicitly cleaned up.
+  """
+
+  use Commanded.Event.Handler,
+    application: Trento.Commanded,
+    name: "saptune_status_update_event_handler"
+
+  import Ecto.Query
+
+  alias Trento.Databases.Events.{
+    DatabaseDeregistered,
+    DatabaseInstanceDeregistered,
+    DatabaseInstanceRegistered,
+    DatabaseRestored
+  }
+
+  alias Trento.SapSystems.Events.{
+    ApplicationInstanceDeregistered,
+    ApplicationInstanceRegistered,
+    SapSystemDeregistered,
+    SapSystemRestored
+  }
+
+  alias Trento.Hosts
+  alias Trento.Hosts.Commands.UpdateSaptuneStatus
+  alias Trento.Hosts.Projections.HostReadModel
+
+  alias Trento.Databases.Projections.DatabaseInstanceReadModel
+  alias Trento.SapSystems.Projections.ApplicationInstanceReadModel
+
+  alias Trento.Repo
+
+  def handle(
+        %DatabaseInstanceRegistered{
+          host_id: host_id
+        },
+        _
+      ) do
+    host_id
+    |> Hosts.get_host_by_id()
+    |> maybe_dispatch_update_saptune_statue(true)
+  end
+
+  def handle(
+        %ApplicationInstanceRegistered{
+          host_id: host_id
+        },
+        _
+      ) do
+    host_id
+    |> Hosts.get_host_by_id()
+    |> maybe_dispatch_update_saptune_statue(true)
+  end
+
+  def handle(
+        %DatabaseInstanceDeregistered{
+          host_id: host_id,
+          instance_number: instance_number
+        },
+        _
+      ) do
+    host_id
+    |> Hosts.get_host_by_id()
+    |> handle_deregistered([instance_number])
+  end
+
+  def handle(
+        %ApplicationInstanceDeregistered{
+          host_id: host_id,
+          instance_number: instance_number
+        },
+        _
+      ) do
+    host_id
+    |> Hosts.get_host_by_id()
+    |> handle_deregistered([instance_number])
+  end
+
+  def handle(
+        %DatabaseDeregistered{
+          database_id: database_id
+        },
+        _
+      ) do
+    DatabaseInstanceReadModel
+    |> where([d], d.database_id == ^database_id)
+    |> Repo.all()
+    |> handle_instances_deregistered(false)
+  end
+
+  def handle(
+        %SapSystemDeregistered{
+          sap_system_id: sap_system_id
+        },
+        _
+      ) do
+    ApplicationInstanceReadModel
+    |> where([s], s.sap_system_id == ^sap_system_id)
+    |> Repo.all()
+    |> handle_instances_deregistered(false)
+  end
+
+  def handle(
+        %DatabaseRestored{
+          database_id: database_id
+        },
+        _
+      ) do
+    DatabaseInstanceReadModel
+    |> where([d], d.database_id == ^database_id)
+    |> Repo.all()
+    |> handle_instances_deregistered(true)
+  end
+
+  def handle(
+        %SapSystemRestored{
+          sap_system_id: sap_system_id
+        },
+        _
+      ) do
+    ApplicationInstanceReadModel
+    |> where([s], s.sap_system_id == ^sap_system_id)
+    |> Repo.all()
+    |> handle_instances_deregistered(true)
+  end
+
+  defp maybe_dispatch_update_saptune_statue(nil, _), do: :ok
```

Comment:
I was imagining statues being thrown all over the places ehehe
```suggestion
  defp maybe_dispatch_update_saptune_status(nil, _), do: :ok
```